packages/react-hooks/use-local-storage/useLocalStorage.ts

```diff
@@ -0,0 +1,48 @@
+import { Dispatch, SetStateAction, useCallback, useState } from "react";
+
+const storeValue = <T>(key: string, value: T): void => {
+  if (!window) return;
+
+  const valueSerialized = JSON.stringify(value);
+  localStorage.setItem(key, valueSerialized);
+};
+
+const getInitialStoredValue = <T>(key: string, initialValue: T): T => {
+  if (!window) return initialValue;
+
+  const storedValue = localStorage.getItem(key);
+  if (storedValue !== null) return JSON.parse(storedValue);
+  else {
+    storeValue(key, initialValue);
+    return initialValue;
+  }
+};
+
+/**
+ * Persist the state with local storage so that it remains after a page refresh. This can be useful
+ * for a dark theme or to record session information. This hook is used in the same way as useState
+ * except that you must pass the storage key in the 1st parameter. If the window object is not
+ * present (as in SSR), useLocalStorage() will return the default value.
+ *
+ * @param key - Key to use in local storage to store value
+ */
+const useLocalStorage = <T>(key: string, initialValue: T): [T, Dispatch<SetStateAction<T>>] | [T] => {
+  const [state, setState] = useState(getInitialStoredValue(key, initialValue));
```

Comment:
maybe we can change `state` to `storedValue` and `setState` to `setStoredValue`? `state` is a really generic word

Reply:
also, consider moving the function `getStoredValue` inside the hook so you can do

```ts
const getStoredValue = () => {
    // use key and initialValue here
}

const [storevValue, setStoredValue] = useState(getStoredValue)
```